### PR TITLE
Use ps.Map for Counters and Sets, remove Metadata in favour of Latest.

### PR DESCRIPTION
**Pull request by tomwilkie**
Part of #502, helps towards #812 & #854.

### CODE REVIEW

**Comment by tomwilkie** [comment]
This only shows a small improvement vs master.

This change:
```
Toms-MacBook-Pro:render twilkie$ go test -bench .
PASS
BenchmarkEndpointRender-8                	   10000	    120894 ns/op	   46217 B/op	     655 allocs/op
BenchmarkEndpointStats-8                 	100000000	        18.9 ns/op       0 B/op	       0 allocs/op
BenchmarkProcessRender-8                 	   10000	    229275 ns/op	   87324 B/op	    1166 allocs/op
BenchmarkProcessStats-8                  	20000000	        64.5 ns/op       0 B/op	       0 allocs/op
BenchmarkProcessWithContainerNameRender-8	    5000	    259938 ns/op	   96385 B/op	    1310 allocs/op
BenchmarkProcessWithContainerNameStats-8 	20000000	        75.3 ns/op       0 B/op	       0 allocs/op
BenchmarkProcessNameRender-8             	    5000	    289304 ns/op	  110414 B/op	    1414 allocs/op
BenchmarkProcessNameStats-8              	100000000	        20.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerRender-8               	    3000	    426344 ns/op	  149200 B/op	    2037 allocs/op
BenchmarkContainerStats-8                	    3000	    355616 ns/op	  128917 B/op	    1825 allocs/op
BenchmarkContainerWithImageNameRender-8  	    3000	    435100 ns/op	  160829 B/op	    2170 allocs/op
BenchmarkContainerWithImageNameStats-8   	    3000	    350997 ns/op	  128936 B/op	    1825 allocs/op
BenchmarkContainerImageRender-8          	    2000	    523001 ns/op	  197214 B/op	    2494 allocs/op
BenchmarkContainerImageStats-8           	100000000	        19.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerHostnameRender-8       	    3000	    423759 ns/op	  156442 B/op	    2131 allocs/op
BenchmarkContainerHostnameStats-8        	100000000	        18.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkHostRender-8                    	   10000	    125865 ns/op	   53674 B/op	     591 allocs/op
BenchmarkHostStats-8                     	20000000	        64.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodRender-8                     	    2000	    502690 ns/op	  189855 B/op	    2402 allocs/op
BenchmarkPodStats-8                      	100000000	        19.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodServiceRender-8              	    2000	    596074 ns/op	  226953 B/op	    2728 allocs/op
BenchmarkPodServiceStats-8               	100000000	        18.9 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/weaveworks/scope/render	33.525s
```

master:
```
Toms-MacBook-Pro:render twilkie$ go test -bench .
PASS
BenchmarkEndpointRender-8                	   10000	    136654 ns/op	   52423 B/op	     773 allocs/op
BenchmarkEndpointStats-8                 	100000000	        19.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessRender-8                 	    5000	    252663 ns/op	   97502 B/op	    1372 allocs/op
BenchmarkProcessStats-8                  	20000000	        63.9 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessWithContainerNameRender-8	    5000	    284485 ns/op	  107775 B/op	    1540 allocs/op
BenchmarkProcessWithContainerNameStats-8 	20000000	        73.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessNameRender-8             	    5000	    322436 ns/op	  121794 B/op	    1639 allocs/op
BenchmarkProcessNameStats-8              	100000000	        19.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerRender-8               	    3000	    438796 ns/op	  167064 B/op	    2390 allocs/op
BenchmarkContainerStats-8                	    3000	    381435 ns/op	  145151 B/op	    2141 allocs/op
BenchmarkContainerWithImageNameRender-8  	    3000	    470924 ns/op	  179896 B/op	    2546 allocs/op
BenchmarkContainerWithImageNameStats-8   	    3000	    388080 ns/op	  145163 B/op	    2142 allocs/op
BenchmarkContainerImageRender-8          	    2000	    559187 ns/op	  218356 B/op	    2908 allocs/op
BenchmarkContainerImageStats-8           	100000000	        19.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerHostnameRender-8       	    3000	    458770 ns/op	  174065 B/op	    2484 allocs/op
BenchmarkContainerHostnameStats-8        	100000000	        19.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkHostRender-8                    	   10000	    145496 ns/op	   61871 B/op	     709 allocs/op
BenchmarkHostStats-8                     	20000000	        69.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodRender-8                     	    2000	    554012 ns/op	  210833 B/op	    2802 allocs/op
BenchmarkPodStats-8                      	100000000	        20.7 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodServiceRender-8              	    2000	    693266 ns/op	  249112 B/op	    3152 allocs/op
BenchmarkPodServiceStats-8               	50000000	        20.1 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/weaveworks/scope/render	34.110s
```

**Comment by paulbellamy** [comment]
rebased on master after #752 merge

**Comment by paulbellamy** [comment]
After rebasing, with a slow report, it looks to give a ~20% decrease in render times.

**Comment by tomwilkie** [comment]
Sweet!

I think next steps are doing this for metadata (or moving all the uses of
metadata to latest) and doing this for controls.

I also want to figure out a way of measuring the cpu cost on a per path
basis, maybe by just measuring latency of each http request?

**Comment by tomwilkie** [comment]
Here's the current performance after rebase:

```
Toms-MacBook-Pro:render twilkie$ go test -bench . -- -bench-report-file report.json 
PASS
BenchmarkEndpointRender-8                	   10000	    119374 ns/op	   46749 B/op	     638 allocs/op
BenchmarkEndpointStats-8                 	100000000	        19.3 ns/op       0 B/op	       0 allocs/op
BenchmarkProcessRender-8                 	    5000	    270373 ns/op	   98960 B/op	    1211 allocs/op
BenchmarkProcessStats-8                  	20000000	        68.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessWithContainerNameRender-8	    5000	    278818 ns/op	  114341 B/op	    1382 allocs/op
BenchmarkProcessWithContainerNameStats-8 	20000000	        76.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessNameRender-8             	    5000	    309158 ns/op	  127276 B/op	    1475 allocs/op
BenchmarkProcessNameStats-8              	100000000	        19.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerRender-8               	    3000	    446979 ns/op	  186119 B/op	    2183 allocs/op
BenchmarkContainerStats-8                	    5000	    371770 ns/op	  153076 B/op	    1944 allocs/op
BenchmarkContainerWithImageNameRender-8  	    3000	    486974 ns/op	  199662 B/op	    2328 allocs/op
BenchmarkContainerWithImageNameStats-8   	    3000	    376005 ns/op	  153085 B/op	    1944 allocs/op
BenchmarkContainerImageRender-8          	    2000	    600634 ns/op	  255532 B/op	    2703 allocs/op
BenchmarkContainerImageStats-8           	100000000	        21.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerHostnameRender-8       	    2000	    579079 ns/op	  193635 B/op	    2277 allocs/op
BenchmarkContainerHostnameStats-8        	100000000	        20.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkHostRender-8                    	    2000	    878810 ns/op	  329441 B/op	    3489 allocs/op
BenchmarkHostStats-8                     	10000000	       139 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodRender-8                     	    2000	    590655 ns/op	  235847 B/op	    2587 allocs/op
BenchmarkPodStats-8                      	100000000	        20.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodServiceRender-8              	    2000	    710204 ns/op	  291872 B/op	    2968 allocs/op
BenchmarkPodServiceStats-8               	100000000	        20.5 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/weaveworks/scope/render	39.250s
```

**Comment by tomwilkie** [comment]
After rebase:

```
Toms-MacBook-Pro:render twilkie$ go test -bench . -- -bench-report-file report.json 
PASS
BenchmarkEndpointRender-8                	   10000	    114758 ns/op	   46583 B/op	     636 allocs/op
BenchmarkEndpointStats-8                 	100000000	        20.6 ns/op       0 B/op	       0 allocs/op
BenchmarkProcessRender-8                 	   10000	    244565 ns/op	   98794 B/op	    1209 allocs/op
BenchmarkProcessStats-8                  	20000000	        69.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessWithContainerNameRender-8	    5000	    276988 ns/op	  114176 B/op	    1380 allocs/op
BenchmarkProcessWithContainerNameStats-8 	20000000	        78.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessNameRender-8             	    5000	    298939 ns/op	  127116 B/op	    1473 allocs/op
BenchmarkProcessNameStats-8              	100000000	        20.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerRender-8               	    3000	    446539 ns/op	  185989 B/op	    2181 allocs/op
BenchmarkContainerStats-8                	    3000	    375755 ns/op	  152921 B/op	    1942 allocs/op
BenchmarkContainerWithImageNameRender-8  	    3000	    481497 ns/op	  199505 B/op	    2327 allocs/op
BenchmarkContainerWithImageNameStats-8   	    3000	    394297 ns/op	  152895 B/op	    1942 allocs/op
BenchmarkContainerImageRender-8          	    2000	    704382 ns/op	  255313 B/op	    2701 allocs/op
BenchmarkContainerImageStats-8           	100000000	        20.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerHostnameRender-8       	    3000	    489741 ns/op	  193500 B/op	    2275 allocs/op
BenchmarkContainerHostnameStats-8        	100000000	        21.5 ns/op	       0 B/op	       0 allocs/op
BenchmarkHostRender-8                    	    2000	    776683 ns/op	  330908 B/op	    3485 allocs/op
BenchmarkHostStats-8                     	10000000	       135 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodRender-8                     	    2000	    565940 ns/op	  235692 B/op	    2585 allocs/op
BenchmarkPodStats-8                      	100000000	        20.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodServiceRender-8              	    2000	    687174 ns/op	  291711 B/op	    2967 allocs/op
BenchmarkPodServiceStats-8               	100000000	        20.2 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/weaveworks/scope/render	37.441s
```

**Comment by tomwilkie** [comment]
Removing Metadata seems to have made a bit of a difference:
```
Toms-MacBook-Pro:render twilkie$ go test -bench . -- -bench-report-file report.json 
PASS
BenchmarkEndpointRender-8                	   20000	     93756 ns/op	   36560 B/op	     537 allocs/op
BenchmarkEndpointStats-8                 	100000000	        20.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessRender-8                 	   10000	    199651 ns/op	   81129 B/op	    1048 allocs/op
BenchmarkProcessStats-8                  	20000000	        66.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessWithContainerNameRender-8	   10000	    252394 ns/op	   93416 B/op	    1197 allocs/op
BenchmarkProcessWithContainerNameStats-8 	20000000	        82.0 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessNameRender-8             	    5000	    280984 ns/op	  107636 B/op	    1299 allocs/op
BenchmarkProcessNameStats-8              	100000000	        20.2 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerRender-8               	    2000	    539458 ns/op	  183885 B/op	    2268 allocs/op
BenchmarkContainerStats-8                	    5000	    380214 ns/op	  154729 B/op	    2027 allocs/op
BenchmarkContainerWithImageNameRender-8  	    3000	    467436 ns/op	  191281 B/op	    2388 allocs/op
BenchmarkContainerWithImageNameStats-8   	    3000	    379499 ns/op	  154744 B/op	    2027 allocs/op
BenchmarkContainerImageRender-8          	    2000	    533797 ns/op	  232635 B/op	    2728 allocs/op
BenchmarkContainerImageStats-8           	100000000	        19.4 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerHostnameRender-8       	    3000	    449177 ns/op	  191396 B/op	    2362 allocs/op
BenchmarkContainerHostnameStats-8        	100000000	        19.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkHostRender-8                    	    2000	    698756 ns/op	  296103 B/op	    3438 allocs/op
BenchmarkHostStats-8                     	10000000	       127 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodRender-8                     	    2000	    525026 ns/op	  224886 B/op	    2661 allocs/op
BenchmarkPodStats-8                      	100000000	        19.8 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodServiceRender-8              	    2000	    669774 ns/op	  265491 B/op	    3021 allocs/op
BenchmarkPodServiceStats-8               	100000000	        19.7 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/weaveworks/scope/render	40.533s
```

In particular, HostRender (our most expensive) has gone form 776683 -> 698756, a 10% win.

**Comment by tomwilkie** [comment]
On master HostRendered is about 884358, so this is about 20% performance improvement - ie not enough.

**Comment by tomwilkie** [comment]
Another improvement (about 30% now)

```
Toms-MacBook-Pro:render twilkie$ go test -bench . -- -bench-report-file report.json 
PASS
BenchmarkEndpointRender-8                	   20000	     89061 ns/op	   26108 B/op	     455 allocs/op
BenchmarkEndpointStats-8                 	50000000	        23.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessRender-8                 	   10000	    178029 ns/op	   65252 B/op	     905 allocs/op
BenchmarkProcessStats-8                  	20000000	        69.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessWithContainerNameRender-8	   10000	    220903 ns/op	   77279 B/op	    1047 allocs/op
BenchmarkProcessWithContainerNameStats-8 	20000000	        79.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkProcessNameRender-8             	   10000	    234506 ns/op	   88773 B/op	    1132 allocs/op
BenchmarkProcessNameStats-8              	100000000	        19.6 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerRender-8               	    3000	    378241 ns/op	  148396 B/op	    1953 allocs/op
BenchmarkContainerStats-8                	    5000	    322733 ns/op	  120321 B/op	    1728 allocs/op
BenchmarkContainerWithImageNameRender-8  	    3000	    408063 ns/op	  155543 B/op	    2067 allocs/op
BenchmarkContainerWithImageNameStats-8   	    5000	    321084 ns/op	  120331 B/op	    1728 allocs/op
BenchmarkContainerImageRender-8          	    3000	    479331 ns/op	  193134 B/op	    2369 allocs/op
BenchmarkContainerImageStats-8           	100000000	        19.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkContainerHostnameRender-8       	    3000	    403683 ns/op	  155901 B/op	    2046 allocs/op
BenchmarkContainerHostnameStats-8        	100000000	        19.3 ns/op	       0 B/op	       0 allocs/op
BenchmarkHostRender-8                    	    2000	    613970 ns/op	  250425 B/op	    3013 allocs/op
BenchmarkHostStats-8                     	10000000	       127 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodRender-8                     	    3000	    463328 ns/op	  186217 B/op	    2310 allocs/op
BenchmarkPodStats-8                      	100000000	        19.1 ns/op	       0 B/op	       0 allocs/op
BenchmarkPodServiceRender-8              	    2000	    549846 ns/op	  223603 B/op	    2639 allocs/op
BenchmarkPodServiceStats-8               	100000000	        19.1 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/weaveworks/scope/render	38.055s
```

**Comment by paulbellamy** [comment]
This panics running the benchmarks against report.json.

**Comment by paulbellamy** [comment]
Basically just some more comments on tests, then LGTM.

Added a minor optimisation that gains another ~10%.

File: report/edge_metadatas_internal_test.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+		})
+	have := EmptyEdgeMetadatas.
+		Add("foo",
+		EdgeMetadata{
+			EgressPacketCount: newu64(3),
+		})
+	if !reflect.DeepEqual(want, have) {
+		t.Errorf(test.Diff(want, have))
+	}
+}
```
Comment (paulbellamy) [inline report/edge_metadatas_internal_test.go]:
```Go
func (c Counters) DeepEquals(i interface{}) bool {
  return true
}
```

Comment (tomwilkie) [inline report/edge_metadatas_internal_test.go]: Fair point, fixed.

**Pull request merged**